UPD: Unified_use_hightlight.js

rgaricano opened the pull request:
### UPD:   Unifying the duplicate highlight.js usage to reduce bundle size.

# Changelog Entry

### Description

-  To solve the bundle duplication problem where both CodeMirror's native language packages and highlight.js were being included separately.

### Changed

- Configure CodeMirror without language-specific packages.
- Created custom CodeMirror HighlightJSExtension.js, to use highlight.js as the highlighter for CodeMirror's syntax highlighting.
- Removed unused libs due to changes from package.json & package-lock.json:
    - "codemirror-lang-elixir": "^4.0.0"
    - "codemirror-lang-hcl": "^0.1.0"
_____
### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

rgaricano committed:
Create HighlightJSExtension.js

UPD-Unified_use_hightlight.js

Create HighlightJSExtension.js

Custom CodeMirror extension that bridges highlight.js syntax highlighting with CodeMirror's decoration system.

rgaricano committed:
UPD-Unified_use_hightlight.js

 UPD-Unified_use_hightlight.js

Update CodeEditor.svelte

Changes necessaries to solve the bundle duplication problem where both CodeMirror's native language packages and highlight.js were being included separately

rgaricano committed:
UPD-Unified_use_hightlight.js

Update of package.json

Removed Unused libs after changes to solve the bundle duplication problem where both CodeMirror's native language packages and highlight.js were being included separately

rgaricano committed:
UPD-Unified_use_hightlight.js

UPD-Unified_use_hightlight.js

Update of package-lock.json

 Removed Unused libs after changes to solve the bundle duplication problem where both CodeMirror's native language packages and highlight.js were being included separately

ShirasawaSama commented:
I will recheck to confirm that the duplicate highlight.js usage have been correctly removed.

> _But I briefly looked at the commit, did you rerun `npm install --force`, it seems to me that the `package-lock.json` file was changed manually?_

ShirasawaSama commented:
<img width="2698" height="830" alt="QQ_1757266426565" src="https://github.com/user-attachments/assets/1b6ceb78-d0b6-4e28-a186-f1c3dd131926" />

The current code appears to be failing compilation. I recommend converting it to draft to prevent accidental merging.

<img width="2088" height="868" alt="QQ_1757266504965" src="https://github.com/user-attachments/assets/5c6288e4-347f-4cfe-b877-435b1c236678" />

rgaricano committed:
Update package.json

error corrected

rgaricano commented:
Thanks, @ShirasawaSama ,
 Yes, I removed wrong lib when made the PR (I edited on my github repo directly).
the libraries to be removed are:
		"codemirror-lang-elixir": "^4.0.0",
		"codemirror-lang-hcl": "^0.1.0",

ShirasawaSama commented on src/lib/components/common/CodeEditor.svelte:
```diff
@@ -6,8 +6,8 @@
 
-	import { indentUnit, LanguageDescription } from '@codemirror/language';
-	import { languages } from '@codemirror/language-data';
+	import { highlightJSExtension } from './CodeEditor/HighlightJSExtension.js';
```
should be `'./HighlightJSExtension.js'`

ShirasawaSama commented:
On the other hand, I think it would be better if language support for dynamic import?

To me, I think default to loading language support for JavaScript, TypeScript, Python, HTML, and CSS is necessary.

tjbck committed:
refac/enh: commands ui

tjbck committed:
refac: deprecate textarea input

tjbck committed:
refac: channel input

rgaricano committed:
Update CodeEditor.svelte

Added modifications to use compartments for HightlightJSExtension.js
to update code hightlight syntax immediately

rgaricano committed:
Update HighlightJSExtension.js

rgaricano committed:
Update package-lock.json

rgaricano commented:
@ShirasawaSama,
I think that all is fixed.
i had to add HCL language definition to HightlightJSExtension.js to mantain actual languages supported.

rgaricano committed:
Update RichTextInput.svelte

Adaptation for HightLight

ShirasawaSama commented:
Great job! I will have a code review.

<img width="2386" height="2052" alt="image" src="https://github.com/user-attachments/assets/d0cd69f0-ec81-411b-9f11-4670b222cfdd" />

ShirasawaSama commented on package.json:
```diff
@@ -138,6 +137,7 @@
 		"uuid": "^9.0.1",
 		"vite-plugin-static-copy": "^2.2.0",
 		"y-prosemirror": "^1.3.7",
+		"y-protocols":"^1.0.6",		
```
It seems that this dependency was not used in this code? Can it be removed?

rgaricano replied:
Yes, but when I build the code always have errors with it, I made a PR about:
https://github.com/open-webui/open-webui/pull/16677
It seem that it's a y-prosemirror dependent: https://github.com/open-webui/open-webui/pull/16677#issuecomment-3197929163

I'don't know why the error on my end ??

I'll remove both (visualizer and y-protocols)

ShirasawaSama commented on src/lib/components/common/RichTextInput.svelte:
```diff
@@ -152,15 +152,44 @@
+	  }
+
+	  try {
+    	const languageModule = await import(`highlight.js/lib/languages/${lang}`);
```
Dynamic import seems to be ineffective because all language dependencies have already been introduced in `highlight.js/lib/index.js`

<img width="1130" height="916" alt="image" src="https://github.com/user-attachments/assets/c5cb514b-14c5-4332-b38c-ee309f85f9f1" />

From the results, it can also be seen that the parsers for each language were not individually packaged into small files:

<img width="448" height="1314" alt="image" src="https://github.com/user-attachments/assets/1954594a-b48d-4a42-be96-4b621c917d9b" />

I think it may not be possible to directly import the 'index.js' file or it can only be imported from the 'highlight.js/es' directory

rgaricano replied:
I'm trying with a separate language loader.

rgaricano committed:
Update CodeEditor.svelte

rgaricano committed:
Update RichTextInput.svelte

rgaricano committed:
Update CodeBlock.svelte

rgaricano committed:
Add files via upload

rgaricano committed:
Update index.ts

rgaricano committed:
Update package.json

rgaricano committed:
Update package-lock.json

ShirasawaSama commented on src/lib/components/common/RichTextInput.svelte:
```diff
@@ -983,7 +993,7 @@
 					link: link
 				}),
-				Placeholder.configure({ placeholder: () => _placeholder }),
+				Placeholder.configure({ placeholder }),
```
This commit appears to contain other commits from the dev branch and may not have been rebased correctly.

We should not use `merge` to incorporate changes from the latest dev branch. Instead, use `git rebase dev` or `git rebase upstream/dev` (if your upstream source is the original repository).

rgaricano committed:
Update RichTextInput.svelte

Isaeiah committed:
Create HighlightJSExtension.js

UPD-Unified_use_hightlight.js

Create HighlightJSExtension.js

Custom CodeMirror extension that bridges highlight.js syntax highlighting with CodeMirror's decoration system.
 UPD-Unified_use_hightlight.js

 UPD-Unified_use_hightlight.js

Update CodeEditor.svelte

Changes necessaries to solve the bundle duplication problem where both CodeMirror's native language packages and highlight.js were being included separately
UPD-Unified_use_hightlight.js

Update of package.json

Removed Unused libs after changes to solve the bundle duplication problem where both CodeMirror's native language packages and highlight.js were being included separately
UPD-Unified_use_hightlight.js

UPD-Unified_use_hightlight.js

Update of package-lock.json

 Removed Unused libs after changes to solve the bundle duplication problem where both CodeMirror's native language packages and highlight.js were being included separately
Update package.json

error corrected
Update CodeEditor.svelte

Added modifications to use compartments for HightlightJSExtension.js
to update code hightlight syntax immediately
Update HighlightJSExtension.js
Update package-lock.json

Update RichTextInput.svelte

Adaptation for HightLight

refac/enh: commands ui

refac: deprecate textarea input

refac: channel input

Update CodeEditor.svelte
Update RichTextInput.svelte

Update CodeBlock.svelte

Add files via upload
Update index.ts
Update package.json

Update package-lock.json

Update RichTextInput.svelte

rgaricano committed:
Merge branch 'UPD-Unified_use_hightlight.js' of https://github.com/rgaricano/open-webui into UPD-Unified_use_hightlight.js

rgaricano committed:
Revised

rgaricano committed:
Delete backend/selfsigned.key

rgaricano committed:
update packages

rgaricano committed:
Merge branch 'UPD-Unified_use_hightlight.js' of https://github.com/rgaricano/open-webui into UPD-Unified_use_hightlight.js

rgaricano committed:
Update highlightLanguageLoader.js

Format

rgaricano committed:
Update highlightLanguageLoader.js

fix }

ShirasawaSama commented:
<img width="878" height="682" alt="QQ_1758279657001" src="https://github.com/user-attachments/assets/cccc1250-3012-4ceb-b504-deb00424eac2" />

rgaricano commented:
No, in that file must to be only one line changed, I revised and seem that I make something wrong when rebasing.

Now it seem that all is ok.
(I think that package.json & package-lock.json are ok, but as I have to edit it to add "y-protocols":"^1.0.6" for build and then remove for pull,...,  only this lib have to be removed:
- "@codemirror/lang-javascript": "^6.2.2",  
- "@codemirror/lang-python": "^6.1.6",  
- "@codemirror/language-data": "^6.5.1",  
- "codemirror-lang-elixir": "^4.0.0",  
- "codemirror-lang-hcl": "^0.1.0",
)

ShirasawaSama commented:
```js
import { defineConfig, type Plugin } from 'vite';
import path from 'path';
import fs from 'fs/promises';
const HighlightJSLanguagesPlugin = (): Plugin => {
	const virtualModuleId = 'virtual:highlightjs-languages';
	const resolvedVirtualModuleId = '\0' + virtualModuleId;

	return {
		name: 'highlightjs-languages-plugin',
		resolveId(id) {
			if (id === virtualModuleId) {
				return resolvedVirtualModuleId;
			}
		},
		async load(id) {
			if (id !== resolvedVirtualModuleId) {
				return;
			}
			const languagesPath = path.join(import.meta.resolve('highlight.js').replace('file:', ''), '../languages');
			const languages = (await fs.readdir(languagesPath)).filter(file => !file.endsWith('.js.js')).map(file => file.replace(/\.js$/, ''));

			const { default: hljs } = await import('highlight.js');
			const generateLanguageConfig = (language: string) => `() => import(${JSON.stringify(path.join(languagesPath, language))})`;
			const kvPairs = languages.map(language => [language, generateLanguageConfig(language)]);
			languages.forEach(language => {
				const config = hljs.getLanguage(language);
				if (config?.aliases?.length) {
					config.aliases.forEach(alias => kvPairs.push([alias, generateLanguageConfig(language)]));
				}
			});

			return `export default { ${kvPairs.map(([key, value]) => `'${key}': ${value}`).join(', ')} }`
		},
	}
}
```

<img width="1762" height="670" alt="QQ_1758350972656" src="https://github.com/user-attachments/assets/0ec596d5-15c0-46d5-970b-30c8b806ac5d" />

ShirasawaSama commented:
> I think that isn't a good idea hardcode the highlight languages to import, if its are changed it's necessary to change the code (& maybe some users use other lib versions in its devs). I don't see any problem in the warning.

<img width="1014" height="134" alt="QQ_1758351163244" src="https://github.com/user-attachments/assets/62948efa-410e-4427-aa85-1fbc768dfc05" />
